File: exercices/solutions/pi_collect.c
#include <mpi.h>
#include <math.h>
#include <stdio.h>

int main(int argc, char *argv[])
{
    int rank, size, n, debut, fin, k;
    double somme, pi, t1, t2;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    do {
        if (rank == 0) {
            printf("Enter the number of intervals (0 quits) : ");
            fflush(stdout);
            scanf("%d", &n);
            if (n > 0) printf("-> %d iterations\n", n); else printf("done\n");
            t1 = MPI_Wtime();
        }

        MPI_Bcast(&n, 1, MPI_INT, 0, MPI_COMM_WORLD);
        if (n == 0) break;

        debut = rank * (long)n / size;
        fin = (rank + 1) * (long)n / size;

        somme = 0.0;
        for (k = fin - 1; k >= debut; --k) {
            somme += (4.0 - 8.0 * (k % 2)) / (2.0 * k + 1);
        }

        MPI_Reduce(&somme, &pi, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

        if (rank == 0) {
            t2 = MPI_Wtime();
            printf("PI is approximately %.16f, difference is %.16f\n",
                   pi, fabs(pi - M_PI));
            printf("Time = %.6f sec\n", t2 - t1);
        }
    } while (1);

    MPI_Finalize();
    return 0;
}

File: exercices/solutions/pi_collect.py
#!/usr/bin/env python

from mpi4py import MPI  # MPI.Init() implicite
import math
import sys


def main():
    """
    Programme principal
    """

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()
    n = None

    while True:
        if rank == 0:
            n = int(input('Enter the number of intervals (0 quits): '))
            print(f'-> {n} iterations' if n > 0 else 'done')
            t1 = MPI.Wtime()

        n = comm.bcast(n, 0)
        if n == 0:
            break

        debut = rank * n // size
        fin = (rank + 1) * n // size

        somme = 0.0;
        for k in range(fin - 1, debut - 1, -1):
            somme += (4.0 - 8.0 * (k % 2)) / (2.0 * k + 1)

        pi = comm.reduce(somme, MPI.SUM, 0)

        if rank == 0:
            t2 = MPI.Wtime()
            print(f'PI is approximately {pi:.16f},',
                  f'difference is {abs(pi - math.pi):.16f}')
            print(f'Time = {t2 - t1:.6f} sec')

    MPI.Finalize()


if __name__ == '__main__':
    main()
